subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/internal/precompiled/PreCompiledScript.java

```diff
@@ -0,0 +1,171 @@
+        private final Class<?> targetClass;
+        private final String fileExtension;
+
+        Type(Class<?> targetClass, String fileExtension) {
+            this.targetClass = targetClass;
+            this.fileExtension = fileExtension;
+        }
+
+        static Type getType(String fileName) {
+            if (fileName.endsWith(SETTINGS.fileExtension)) {
+                return SETTINGS;
+            }
+            if (fileName.endsWith(INIT.fileExtension)) {
+                return INIT;
+            }
+            return PROJECT;
+        }
+
+        PluginId toPluginId(String fileName) {
+            return DefaultPluginId.of(fileName.substring(0, fileName.indexOf(fileExtension)));
+        }
+    }
+
+    PreCompiledScript(File scriptFile) {
+        this.scriptSource = new TextResourceScriptSource(new UriTextResource("script", scriptFile));
+        String fileName = scriptFile.getName();
+        this.type = Type.getType(fileName);
+        this.pluginId = type.toPluginId(fileName);
+        this.scriptTarget = new PreCompiledScriptTarget(type == Type.PROJECT);
+    }
+
+    static void filterPluginFiles(PatternFilterable patternFilterable) {
+        patternFilterable.include("**/*" + SCRIPT_PLUGIN_EXTENSION);
+    }
+
+    void declarePlugin(PluginDeclaration pluginDeclaration) {
+        pluginDeclaration.setImplementationClass(getGeneratedPluginClassName());
+        pluginDeclaration.setId(pluginId.getId());
+    }
+
+    String getId() {
+        return pluginId.getId();
+    }
+
+    String getGeneratedPluginClassName() {
+        return toJavaIdentifier(kebabCaseToPascalCase(pluginId.getName())) + "Plugin";
+    }
+
+    String getClassName() {
+        return scriptSource.getClassName();
+    }
+
+    String getPluginsBlockClassName() {
+        return getPluginsBlockSource().getClassName();
+    }
+
+    HashCode getContentHash() {
+        return scriptSource.getResource().getContentHash();
+    }
+
+    ScriptSource getSource() {
+        return scriptSource;
+    }
+
+    ScriptSource getPluginsBlockSource() {
+        return new PluginsBlockSourceWrapper(scriptSource);
+    }
+
+    private static String kebabCaseToPascalCase(String s) {
+        return CaseFormat.LOWER_HYPHEN.to(CaseFormat.UPPER_CAMEL, s);
+    }
+
+    private static String toJavaIdentifier(String s) {
+        StringBuilder sb = new StringBuilder();
+        if (!Character.isJavaIdentifierStart(s.charAt(0))) {
+            sb.append("_");
+        }
+        for (char c : s.toCharArray()) {
+            if (Character.isJavaIdentifierPart(c)) {
+                sb.append(c);
+            } else {
+                sb.append("_");
+            }
+        }
+        return sb.toString();
+    }
+
+    public ScriptTarget getScriptTarget() {
+        return scriptTarget;
+    }
+
+    public String getTargetClassName() {
+        return type.targetClass.getName();
+    }
+
+    private static class PluginsBlockSourceWrapper implements ScriptSource {
```

Comment:
❓ is this new implementation of `ScriptSource` necessary?

Reply:
This one is a bit more tricky. The script is compiled in two phases - plugins block, then body. Using the existing infrastructure, generated artifacts have the same class name and they clash when packaged in a single Jar.
This is more of a workaround to force plugins block class name to be different from the body by prefixing it with `plugins_`. It might be fixable deeper down, however, I wanted to keep this change as non-invasive as possible.